teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md

```diff
@@ -171,12 +201,503 @@ In addition to grouping related resources, Tekton Results provides long term sto
+        message: "Tasks Completed: 4, Skipped: 0"
+        reason: Succeeded
+        status: "True"
+        type: Succeeded
+    startTime: "2020-05-04T02:00:11Z"
+    resources:
+      - apiVersion: v1beta1
+        kind: TaskRun
+        name: triggers-release-nightly-build
+        conditions:
+          - lastTransitionTime: "2020-05-04T02:10:49Z"
+            message: All Steps have completed executing
+            reason: Succeeded
+            status: "True"
+            type: Succeeded
+```
+
+#### Discussion
+
+This approach is extensible to support `PipelineRuns` and other types we may support in
+`PipelineTasks` in addition to `TaskRuns` and `Runs`. However, it's more verbose because
+the type and kind have to be repeated for each `TaskRun` and `Run`. It also makes it more
+difficult for tools and users to parse for specific type of resource when looking at the
+status (compared to separate fields [as proposed](#1-add-minimal-embedded-status).
+
+### Add Minimal Embedded Status - Use Map
+
+Use maps for the new fields, as we do with the existing fields:
+
+```go
+type PipelineRunStatusFields struct {
+	...
+	TaskRuns map[string]*PipelineRunTaskRunStatus `json:"taskRuns,omitempty"`
+	TaskRunsStatuses map[string]*PipelineRunTaskRunMinimalStatus `json:"taskRunsStatuses,omitempty"`
+	Runs map[string]*PipelineRunRunStatus `json:"runs,omitempty"`
+	RunsStatuses map[string]*PipelineRunRunMinimalStatus `json:"runsStatuses,omitempty"`
+	...
+}
+```
+
+Taking the [example above](#example), this would be the `PipelineRun` status:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  ...
+spec:
+    ...
+status:
+    completionTime: "2020-05-04T02:19:14Z"
+    conditions:
+      - lastTransitionTime: "2020-05-04T02:19:14Z"
+        message: "Tasks Completed: 4, Skipped: 0"
+        reason: Succeeded
+        status: "True"
+        type: Succeeded
+    startTime: "2020-05-04T02:00:11Z"
+    taskRunStatus:
+      triggers-release-nightly-build:
+        conditions:
+          - lastTransitionTime: "2020-05-04T02:10:49Z"
+            message: All Steps have completed executing
+            reason: Succeeded
+            status: "True"
+            type: Succeeded
+```
+
+#### Discussion
+
+While this approach is consistent with existing code, it does not comply with the
+[Kubernetes API conventions][subobjects] that recommend against maps.
+The [main problem][why-subobjects] with maps is:
+> The crux of maps is that it isn't clear to the user what "left-hand side strings"
+> are "magic keywords" in the config system/API vs. which are user data.
+
+### Beta API - Use Booleans
+
+We could add a behavior flag - `enable-full-embedded-status` - used to configure
+whether `PipelineRuns` should contain the embedded status or minimal status of
+its `TaskRuns` and `Runs`.
+
+Following our [policy][behavior-flags] on updating behavior flags:
+* The `enable-full-embedded-status` flag will be *true* by default.
```

Comment:
We'll need to trigger the deprecation as soon as we add the flag, so we can start counting the 9 months right away.